Simulator coverage for get_account_{transfers,balances}

batiati opened the pull request:
Implemented assertions based on the expected number of results in variations of filters such as `debits | credits` and timestamp range.

It's not able to fine-check intermediate states such as the expected `account_balance` between transfers, although the path covered is good enough to check the assertions already in place along the code.

matklad commented on src/state_machine/workload.zig:
```diff
@@ -503,6 +522,78 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                .one => account_filter.limit = 1,
+                .all => account_filter.limit = transfer_count,
+                .max => account_filter.limit = batch_size,
+            }
```
maybe `account_filter.limit = switch()`? And maybe add `math.intMax(u32)` case just in case?

matklad commented on src/state_machine/workload.zig:
```diff
@@ -503,6 +522,78 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                else
+                    0;
+                break :transfer_count dr_transfer_count + cr_transfer_count;
+            } else 0;
```
Might be more readable in imperative form:

```zig
            var transfer_count: u32 = 0;
            if (self.auditor.get_account_state(account_filter.account_id)) |account_state| {
                if (account_filter.flags.debits) {
                    transfer_count += account_state.dr_transfer_count;
                }
                if (account_filter.flags.credits) {
                    transfer_count += account_state.cr_transfer_count;
                }
            }
```

matklad commented on src/state_machine/workload.zig:
```diff
@@ -503,6 +522,78 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                .max => account_filter.limit = batch_size,
+            }
+
+            // Apply filter by timestamp only when we the prefix or the suffix of the expected
```
"when we the prefix" doesn't parse for me

matklad commented on src/state_machine/workload.zig:
```diff
@@ -503,6 +522,78 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+        fn build_get_account_filter(self: *const Self, body: []tb.AccountFilter) usize {
+            assert(body.len == 1);
+            const account_filter = &body[0];
+            account_filter.* = std.mem.zeroInit(tb.AccountFilter, .{});
```
zeroInit is a bit of a footgun, as it's not actually zeroInit, but rather default init. Using `mem.zero` is somewhat more reliable. 

Though, in this case I think the best is to go with explicit init:

```zig
account_filter.* = .{ /* code here */ }
```

to make sure we didn't forget any fields

matklad commented on src/state_machine/workload.zig:
```diff
@@ -503,6 +522,78 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+            {
+                if (self.auditor.get_account_state(
+                    account_filter.account_id,
+                )) |account_state| {
```
In this branch, we _should_ have account state, right? THen maybe `.?` instead of an if here, to make it fail loudly in case of some future unrelated change?

batiati replied:
It can be an invalid account.

matklad commented on src/state_machine/workload.zig:
```diff
@@ -503,6 +522,78 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                if (self.auditor.get_account_state(
+                    account_filter.account_id,
+                )) |account_state| {
+                    // We want to exclude at least one result:
```
Hm, the code looks like we are trying to _keep_ at least one result?

batiati replied:
We are excluding the first or the last result (if ascending or descending), filtering by timestamp.

```zig
// We want to exclude at least one result:
```

This comment is indeed outdated :+1: 
It says "at least one", but that's no longer true since we can filter by timestamp only when both flags `debits` and `credits` are set, excluding _exactly one_ result.

matklad commented on src/state_machine/workload.zig:
```diff
@@ -503,6 +522,78 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                .created = true,
+                .debits_must_not_exceed_credits = null,
+                .credits_must_not_exceed_debits = null,
+                .has_transfers = true,
```
maybe `has_transfers = null`? Feels like a good case to cover.

matklad replied:
Or maybe just pick account at random? (so, .created = null as well)

matklad commented on src/state_machine/workload.zig:
```diff
@@ -773,6 +861,204 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                    const cr_transfer_count: u32 = if (account_filter.flags.credits)
+                        account_state.cr_transfer_count
+                    else
+                        0;
```
Consider imperative style, should more concise

matklad commented on src/state_machine/workload.zig:
```diff
@@ -773,6 +861,204 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+
+                const expected_count = @min(account_filter.limit, transfer_count, batch_size);
+                if (account_filter.timestamp_min == 0 and account_filter.timestamp_max == 0) {
+                    assert(result_count == expected_count);
```
Hm, how does this work? It seems that new transfers might be created concurrently to our requests? The fact that the account has this many transfers right now doesn't generally mean that it had this many tranfers at the time the lookup was executed. Am I missing some implicit synchronisation logic here here?

batiati replied:
The validation runs _just after_  the query is executed, so the `transfer_count` is up to date. Whatever the number of transfers it has inserted, the number of matches must be one of the three possible values: `@min(account_filter.limit, transfer_count, batch_size)`.

Synchronization will be a problem when implementing `import_{accounts,transfers}` though.

batiati replied:
Reworked the logic, now we only rely on `transfer_count` when asserting the result, never when populating `FuzzOp`.

matklad replied:
>The validation runs just after the query is executed,

Hm, does it? If I am reading the code correctly, this logic is triggered when the client receives a reply, and a reply message might be delayed by the network.

matklad replied:
:thinking: I think if you need to guarantee timing, you could use replica's on commit hook (in replica.zig):

```
            if (self.event_callback) |hook| hook(self, .committed);
```

Though, I'd say its more valuable to try to keep workfload/validator asynchronous, such that it also works for out-of-process scenarious without much pain.

batiati replied:
> his logic is triggered when the client receives a reply, and a reply message might be delayed by the network.

I'm missing something here. When the `Auditor` validates the balance of an account, if new transfers could change that before the validation, wouldn't it be the same problem?

matklad replied:
Excellent point! I am clearly missing something here as well, let me re-read this whole thing, because this couldn't possible work, and yet evidently it does :rofl:

matklad replied:
Ok, yeah, i indeed missed big chunk of logic! Simulator makes use of `reply_sequence: ReplySequence,` to reorder replies in their _logical_ order before calling callbacks!

matklad commented on src/testing/storage.zig:
```diff
@@ -303,8 +303,10 @@ pub const Storage = struct {
+        storage.next_tick_queue.reset();
+        while (q.pop()) |next_tick| {
             next_tick.callback(next_tick);
         }
```
Could you expand the commit explaining a bit _why_ we do it this way? 

It looks like the comment (and correspondingly the logic) were changed to do _exactly_ the thing that the previous comment said should not be done, so I am wondering why :rofl:  (and also why this specific change uncovered this effect)

batiati replied:
This commit fixed the seed `13070689082331695065`, which causes the simulator to loop indefinitely (not during the liveness check), because we finished the prefetch of invalid filters using `next_tick`.

I want it to be exhaustively fuzzed by the CFO because I'm not 100% confident that this fix reproduces exactly the same behavior of the real Storage `on_next_tick`, but apparently, the previous code didn't either. 🤣

batiati replied:
Reverted the code since it crashed [another test](https://github.com/tigerbeetle/tigerbeetle/actions/runs/9585722422/job/26432244243).
This seed doesn't fail anymore, so let it fuzz for some time!

matklad commented on src/state_machine/workload.zig:
```diff
@@ -773,6 +861,204 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                    transfer.timestamp < timestamp_last
+                else
+                    transfer.timestamp > timestamp_last);
+                timestamp_last = transfer.timestamp;
```
I think in loops we usually call such a tracking variable as `_last`

batiati replied:
The name _last is because it's the last timestamp seen, can be the max or min depending on the sorting order.

matklad replied:
Urgh, sorry, I managed to write exactly the opposite of what I wanted to write. We typically call such variables

```
timestamp_previous
```

not last --- "last" is the absolute, last of all there is. previous refers to the latest iteration of the loop.

matklad commented on src/state_machine/workload.zig:
```diff
@@ -773,6 +861,204 @@ pub fn WorkloadType(comptime AccountingStateMachine: type) type {
+                assert(if (account_filter.flags.reversed)
+                    transfer.timestamp < timestamp_last
+                else
+                    transfer.timestamp > timestamp_last);
```
I'd pull `if` outsied of assert (so, duplciating the assert), as it'd be more readable that way. Also, should the assert fail, with two asserts the failure would immediately point a _which_ of the two cases failed. That's basically the same thing why

```
assert(foo)
assert(bar)
```

is superior to

```
assert(foo && bar)
```

matklad reviewed (commented):
Realized that I don't understand one big picture aspect of this: in `validate_account_filter_result_count`, we compare the number of transfers the account has at a given moment with the amount of results. But why than works? More transfers could get created in the meantime, no?

matklad commented:
removed vopr label until this is rebased to 0.13 :)

batiati committed:
Simulator coverage for get_account_*
Implemented assertions based on the expected number of results in variations of filters such as `debits | credits` and timestamp range.

It's not able to fine-check intermediate states such as the expected `account_balance` between transfers, although the path covered is good enough to check the assertions already in place along the code.

matklad commented:
vopr failure:

```
./zig/zig build  simulator_run -- --lite 3734369335599541400
thread 6028 panic: reached unreachable code
/home/matklad/p/tb/review/zig/lib/std/debug.zig:412:14: 0x1120acd in assert (simulator)
    if (!ok) unreachable; // assertion failure
             ^
/home/matklad/p/tb/review/src/state_machine/workload.zig:1032:27: 0x12085fb in validate_account_filter_result_count (simulator)
                    assert(result_count == @min(account_filter.limit, batch_size));
                          ^
/home/matklad/p/tb/review/src/state_machine/workload.zig:990:54: 0x11d22e2 in on_get_account_balances (simulator)
            self.validate_account_filter_result_count(account_filter, results.len);
                                                     ^
/home/matklad/p/tb/review/src/state_machine/workload.zig:365:70: 0x1172136 in on_reply (simulator)
                .get_account_balances => self.on_get_account_balances(
                                                                     ^
/home/matklad/p/tb/review/src/simulator.zig:829:48: 0x1165aca in on_cluster_reply (simulator)
                    simulator.workload.on_reply(
                                               ^
/home/matklad/p/tb/review/src/testing/cluster.zig:679:37: 0x1162019 in client_on_reply (simulator)
            cluster.on_cluster_reply(cluster, client_index, request_message, reply_message);
                                    ^
/home/matklad/p/tb/review/src/vsr/client.zig:439:34: 0x117ad4d in on_reply (simulator)
                on_reply_callback(self, inflight.message, reply);
                                 ^
/home/matklad/p/tb/review/src/vsr/client.zig:179:44: 0x11254c7 in on_message (simulator)
                .reply => |m| self.on_reply(m),
                                           ^
/home/matklad/p/tb/review/src/testing/cluster/network.zig:317:39: 0x12d475e in deliver_message (simulator)
        target_bus.on_message_callback(target_bus, target_message);
                                      ^
/home/matklad/p/tb/review/src/testing/packet_simulator.zig:403:45: 0x1236d78 in tick (simulator)
                        link_packet.callback(link_packet.packet, path);
                                            ^
/home/matklad/p/tb/review/src/testing/cluster/network.zig:136:38: 0x11f1c0c in tick (simulator)
        network.packet_simulator.tick();
```

batiati committed:
Relaxed assertions for get_account_* result count

matklad commented:
Another one!

```
./zig/zig build -Drelease simulator_run -- --lite 1871810428293586998
thread 235499 panic: reached unreachable code
/home/matklad/p/tb/review/zig/lib/std/debug.zig:412:14: 0x10a598e in validate_account_filter_result_count (simulator)
    if (!ok) unreachable; // assertion failure
             ^
/home/matklad/p/tb/review/src/state_machine/workload.zig:897:49: 0x106f3bc in on_reply (simulator)
            validate_account_filter_result_count(
                                                ^
/home/matklad/p/tb/review/src/simulator.zig:829:48: 0x106009f in on_cluster_reply (simulator)
                    simulator.workload.on_reply(
                                               ^
/home/matklad/p/tb/review/src/vsr/client.zig:439:34: 0x1037db8 in on_message (simulator)
                on_reply_callback(self, inflight.message, reply);
                                 ^
/home/matklad/p/tb/review/src/testing/cluster/network.zig:317:39: 0x10e561d in deliver_message (simulator)
        target_bus.on_message_callback(target_bus, target_message);
                                      ^
/home/matklad/p/tb/review/src/testing/packet_simulator.zig:403:45: 0x106894c in tick (simulator)
                        link_packet.callback(link_packet.packet, path);
                                            ^
/home/matklad/p/tb/review/src/simulator.zig:308:23: 0x1066962 in main (simulator)
        simulator.tick();
                      ^
/home/matklad/p/tb/review/zig/lib/std/start.zig:524:37: 0x10360e5 in posixCallMainAndExit (simulator)
            const result = root.main() catch |err| {
                                    ^
/home/matklad/p/tb/review/zig/lib/std/start.zig:266:5: 0x1035b11 in _start (simulator)
    asm volatile (switch (native_arch) {
```

batiati committed:
Narrowed assertions

batiati committed:
s/_first,_last/_min,_max

batiati committed:
Fix comments
